### PR TITLE
add TEP for credentials filter

### PR DESCRIPTION
**Pull request by Useurmind**
Hi,

while working on a pull request regarding credentials filtering we came to the conclusion that a TEP for this feature would make sense.

https://github.com/tektoncd/pipeline/pull/4837

Therefore I added a TEP for this.

### CODE REVIEW

**Comment by Useurmind** [comment]
What does the linting error mean?

**Comment by lbernick** [comment]
> What does the linting error mean?

This is happening because this TEP also needs to be added to the [readme](https://github.com/tektoncd/community/blob/main/teps/README.md)

File: teps/0125-add-credential-filter-to-entrypoint-logger.md
Comment (lbernick) [inline teps/0125-add-credential-filter-to-entrypoint-logger.md:156]: Thanks for including this example-- I was a bit concerned that the PR wouldn't work for workspaces w/ secrets as workspace bindings but based on this I think it will. We should just make sure this case is covered by tests.

File: teps/0125-add-credential-filter-to-entrypoint-logger.md
Comment (lbernick) [inline teps/0125-add-credential-filter-to-entrypoint-logger.md:66]: Does this mean if e.g. your task queries a kube cluster for some secrets and prints their contents this isn't a case we're trying to address? I think that's fine, just a bit confused about what this means.

Possibly another non goal would be trying to redact secrets in params-- I think we just encourage people to not put secrets into params.

**Review by lbernick — COMMENTED** [review]
Thanks @Useurmind for creating this TEP, it's a big help in understanding how this feature works! This generally LGTM but I'm going to see if any more security-focused members of the org are able to take a look as well.

File: teps/0125-add-credential-filter-to-entrypoint-logger.md
Comment (Useurmind) [reply teps/0125-add-credential-filter-to-entrypoint-logger.md:156]: I have just added this to the test case that I created

File: teps/0125-add-credential-filter-to-entrypoint-logger.md
Comment (Useurmind) [reply teps/0125-add-credential-filter-to-entrypoint-logger.md:66]: I see it similar.

When I for example create an Oauth Token during the run of  a step using some client credentials that are mounted into the pod, the client credentials will be redacted but the Oauth Token will NOT.

Regarding redacting parameters that could be another story.

**Comment by Useurmind** [comment]
> > What does the linting error mean?
> 
> This is happening because this TEP also needs to be added to the [readme](https://github.com/tektoncd/community/blob/main/teps/README.md)

I added it to the readme btw

File: teps/0125-add-credential-filter-to-entrypoint-logger.md
Comment (lbernick) [reply teps/0125-add-credential-filter-to-entrypoint-logger.md:66]: probably worth putting "redacting params" in non goals-- useful but more challenging, and out of scope

**Review by lbernick — APPROVED** [review]

File: teps/0125-add-credential-filter-to-entrypoint-logger.md
Comment (Useurmind) [reply teps/0125-add-credential-filter-to-entrypoint-logger.md:66]: I have added this to the non goal section.

Comment (lbernick) [reply teps/0125-add-credential-filter-to-entrypoint-logger.md:66]: awesome thank you!

File: teps/0125-add-credential-filter-to-entrypoint-logger.md
Comment (wlynch) [inline teps/0125-add-credential-filter-to-entrypoint-logger.md:65]: (non-blocking - just another idea) You may also be able to add additional rules for certain types of well-known secret types (i.e. JWT tokens, GitHub tokens, etc.) similar to https://github.com/zricethezav/gitleaks/blob/master/config/gitleaks.toml

File: teps/0125-add-credential-filter-to-entrypoint-logger.md
Comment (wlynch) [inline teps/0125-add-credential-filter-to-entrypoint-logger.md:122]: I suspect we may run into limit issues if we stuff it into an environment variable - not sure what the max size is, but I suspect it's fairly small (especially if things like https://devblogs.microsoft.com/oldnewthing/20100203-00/?p=15083 still hold true). You may want to use an annotation + the [downwards API](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information) with a file instead. Process is mostly the same, just different fields.

**Review by wlynch — APPROVED** [review]
(lgtm pending env vs annotation)

File: teps/0125-add-credential-filter-to-entrypoint-logger.md
Comment (Useurmind) [reply teps/0125-add-credential-filter-to-entrypoint-logger.md:65]: I would say this is also a non goal of this TEP. ;)

The tool you linked is meant to detect more like accidentially added secrets in git repositories. That can happen quite easily when you worked with something that puts secrets in the working directory and commit everything to git. You can detect those with pattern matching.

When you run pipelines in tekton you provision secrets on purpose to the pipeline, either by ENV, file or tekton parameter.

You could provide a secret accidentially in a non standard way like putting it directly into an env or parameter. But you would not be able to pattern match this kind of stuff, because different people will use different names for those env vars and parameters.

To not blow this TEP up to much I would very much like to keep the focus on the described functionality. Handling parameters and accidental secrets is a very different topic and more complicated as this will probably also require API changes. The implementation already done and the related PR is open for quite a long time now and I want to finish the whole thing. We are all eager to finally see this working. ;)

Comment (lbernick) [reply teps/0125-add-credential-filter-to-entrypoint-logger.md:65]: Thanks @Useurmind, I think stating this as a non goal or future work makes a lot of sense. I appreciate your responsiveness to comments on this TEP and your PR, and I'm sorry your PR has been open for so long-- not the contributor experience we aim for. I'll take another look at it this week.

**Comment by lbernick** [comment]
@vdemeester @XinruZhang are you still interested in reviewing?

File: teps/0125-add-credential-filter-to-entrypoint-logger.md
Comment (wlynch) [reply teps/0125-add-credential-filter-to-entrypoint-logger.md:122]: SGTM! I think annotations + downwards API are probably a better starting point so we don't pollute the environment for running steps, and we can reserve a path in `/tekton` so there can't be naming collisions. I think the 256kb limit is fine - I don't think this data will be particularly large if it's only path identifiers. We can look into other ways to pass data to the entrypoint if/when this becomes a problem.

Comment (Useurmind) [reply teps/0125-add-credential-filter-to-entrypoint-logger.md:119]: Ok I adapted the environment variable design to the downward API volume and documented the alternatives we considered.

File: teps/0125-add-credential-filter-to-entrypoint-logger.md
Comment (Useurmind) [reply teps/0125-add-credential-filter-to-entrypoint-logger.md:65]: I am sad to say this but I will not work on this feature set or its specification atm. because I do not have time for it.
Feel free to extend the TEP regarding these topics. I will continue working on the original pull request and adapt it to use the downward API instead of an environment variable.

It there aren't any more design problems with the solution I would regard this pull request as done.

Comment (wlynch) [reply teps/0125-add-credential-filter-to-entrypoint-logger.md:65]: SGTM - no expectation for you to do any additional implementation work! 🙇 

I'm moreso interested in making sure we leave the door open to add this in the future. +1 to @lbernick's suggestion for making a note of this in a `Future Work`section.

Comment (Useurmind) [reply teps/0125-add-credential-filter-to-entrypoint-logger.md:66]: I have added the section for future work

**Review by wlynch — APPROVED** [review]
/hold

Holding for any last comments from @vdemeester or @lbernick. 
Otherwise LGTM!

/lgtm

File: teps/0125-add-credential-filter-to-entrypoint-logger.md
Comment (Useurmind) [reply teps/0125-add-credential-filter-to-entrypoint-logger.md]: Just realized during implementation that secret locations are container specific. That means we would need one annotation per container, e.g. tekton.dev/secret-locations-<containerName>: "{...}"

**Pull request merged**
